### PR TITLE
Add transaction_timestamp, transaction_date, statement_timestamp, clo…

### LINKED ISSUE #2427: Implement current_timestamp, current_date, and now() correctly
current_timestamp, current_date, and now() return the time when they were processed. In postgres they return the same consistent time within a transaction scope, where the recorded time is a timestamp taken  at the start of the transaction.

We could either implement postgres semantics or some other modernized semantics.

### PR DESCRIPTION
**Pull request by vivekmenezes**
…ck_timestamp.

Fix current_timestamp, current_date, now().

closes #2427

### CODE REVIEW

File: sql/parser/builtins.go [hunk 1]
```diff
@@ -382,8 +382,8 @@ var builtins = map[string][]builtin{
 			types:      typeList{timestampType},
 			returnType: DummyInterval,
 			impure:     true,
```
Comment (petermattis) [inline sql/parser/builtins.go]: I think this function and the other time functions (except `clock_timestamp`) are now pure.

File: sql/testdata/builtin_function [hunk 2]
```diff
@@ -258,13 +258,75 @@ query error syntax error at or near "\("
...
+----
+true
+
+query T
+SELECT statement_timestamp - statement_timestamp = interval '0s'
+----
+true
+
+query error syntax error at or near "\("
+SELECT clock_timestamp()
```
Comment (petermattis) [inline sql/testdata/builtin_function]: Is there a reason you want the function form to not work?

File: sql/parser/sql.y [hunk 3]
```diff
@@ -3576,6 +3592,7 @@ reserved_keyword:
 | CASE
 | CAST
 | CHECK
+| CLOCK_TIMESTAMP
```
Comment (petermattis) [inline sql/parser/sql.y]: These are not reserved keywords in the SQL sense. I think you should be adding these "keywords" to the `type_func_name_keyword` list.

File: sql/executor.go [hunk 4]
```diff
@@ -72,6 +75,8 @@ func (e Executor) Execute(args driver.Request) (driver.Response, int, error) {
 			txn.SetSystemDBTrigger()
 		}
 		planMaker.txn = txn
+		now = time.Unix(0, 0).Add(time.Duration(planMaker.session.UnixTimestamp))
```
Comment (petermattis) [inline sql/executor.go]: `proto.Timestamp.GoTime()` does this differently. I'm not sure if it matters, but it would be good be consistent. For that matter, it might be easier if `session.UnixTimestamp` were a `proto.Timestamp` so that you could just use the `GoTime()` method.

File: sql/testdata/builtin_function [hunk 5]
```diff
@@ -258,13 +258,75 @@ query error syntax error at or near "\("
 SELECT current_date()
 
 query T
-SELECT now() - current_timestamp < interval '10s'
+SELECT now() - transaction_date::timestamp < interval '24h10s'
```
Comment (petermattis) [inline sql/testdata/builtin_function]: Other tests for `now()` exist in `testdata/datetime`. Perhaps put these new datetime function tests there as well.

File: proto/data.pb.go [hunk 6]
```diff
@@ -546,7 +546,6 @@ func (m *NodeList) GetNodes() []int32 {
 // transaction is assigned a random priority. This priority will be
 // used to decide whether a transaction will be aborted during
 // contention.
-// TODO(vivek): Remove parts of Transaction that expose internals.
```
Comment (tamird) [inline proto/data.pb.go]: looks like you edited generated code.

File: sql/executor.go (hunk 4 above)
Comment (tamird) [inline sql/executor.go]: :+1: assuming s/`proto.Timestamp`/`driver.Datum.Timestamp`/

File: sql/testdata/builtin_function (hunk 2 above)
Comment (vivekmenezes) [inline sql/testdata/builtin_function]: No strong reason, we should support all functions like unique_id() without the parenthesis. It's nice to not have the parenthesis, making it look like a variable in SQL. I'll allow the variable use as a function.

File: sql/parser/builtins.go (hunk 1 above)
Comment (vivekmenezes) [inline sql/parser/builtins.go]: So purity is defined within the scope of a statement? I'd like to update the comment.

Comment (tamird) [inline sql/parser/builtins.go]: Yeah, this is a bit ambiguous. If you look at how we use `impure` currently, we only care if the expression is a constant, and we only care about that for the purpose of expression simplification.

I think it would be fair to say that purity is statement-scoped, but the update comment should mention the specifics of how `impure` is accessed.

Comment (vivekmenezes) [inline sql/parser/builtins.go]: I've updated the comment to reflect some of these thoughts. I prefer not mentioning how impure is used and the comments reflect only how a builtin writer would specify it.

File: sql/testdata/builtin_function (hunk 2 above)
Comment (vivekmenezes) [inline sql/testdata/builtin_function]: done

File: sql/parser/sql.y (hunk 3 above)
Comment (vivekmenezes) [inline sql/parser/sql.y]: done

File: sql/executor.go (hunk 4 above)
Comment (vivekmenezes) [inline sql/executor.go]: fixed

File: sql/testdata/builtin_function (hunk 5 above)
Comment (vivekmenezes) [inline sql/testdata/builtin_function]: i can do that once you've approved this change. I'll move this code into the other file as a separate commit in this PR.

File: proto/data.pb.go (hunk 6 above)
Comment (vivekmenezes) [inline proto/data.pb.go]: oops! fixed

File: sql/testdata/datetime [hunk 7]
```diff
@@ -72,3 +72,5 @@ query B
 SELECT now() + '1m'::interval > now(), now() + '1m'::interval >= now()
 ----
 true true
+
+
```
Comment (vivekmenezes) [inline sql/testdata/datetime]: let me take this out

File: sql/session.proto [hunk 8]
```diff
@@ -35,4 +35,7 @@ message Session {
   // Indicates that the above transaction is mutating keys in the
   // SystemDB span.
   optional bool mutates_system_db = 4 [(gogoproto.nullable) = false, (gogoproto.customname) = "MutatesSystemDB"];
+  // Timestamp to be used by SQL in the above transaction. Note: this is not the
+  // transaction timestamp in proto.Transaction above.
+  optional proto.Timestamp unix_timestamp = 5 [(gogoproto.nullable) = false];
```
Comment (petermattis) [inline sql/session.proto]: We usually name this field `timestamp`. I'm not sure why we need to highlight that it is a unix timestamp in the name. The comments in `proto.Timestamp` already indicate that it is based on the unix epoch.

**Comment by petermattis** [comment]
LGTM

File: sql/session.proto (hunk 8 above)
Comment (tamird) [inline sql/session.proto]: this is not the right type. We should use the driver's `Timestamp` here, not `proto`'s/

File: sql/parser/sql.y [hunk 9]
```diff
@@ -2773,15 +2774,15 @@ func_expr_windowless:
...
+  func_or_variable
   {
     $$ = &FuncExpr{Name: &QualifiedName{Base: Name($1)}}
   }
-| CURRENT_TIMESTAMP
+| func_or_variable '(' ')'
   {
     $$ = &FuncExpr{Name: &QualifiedName{Base: Name($1)}}
   }
+| COLLATION FOR '(' a_expr ')' { unimplemented() }
```
Comment (tamird) [inline sql/parser/sql.y]: nit: why move this?

**Comment by tamird** [comment]
LGTM modulo the timestamp using the correct proto message type.

File: sql/executor.go [hunk 10]
```diff
@@ -72,6 +75,8 @@ func (e Executor) Execute(args driver.Request) (driver.Response, int, error) {
 			txn.SetSystemDBTrigger()
 		}
 		planMaker.txn = txn
+		now = planMaker.session.UnixTimestamp.GoTime()
```
Comment (bdarnell) [inline sql/executor.go]: It's weird that after this point `now` is not the current time. This should use a different local variable `txnTimestamp` (which defaults to `now`).

File: sql/parser/builtins.go [hunk 11]
```diff
@@ -399,15 +401,46 @@ var builtins = map[string][]builtin{
...
 			returnType: DummyDate,
+			fn: func(e EvalContext, args DTuple) (Datum, error) {
+				return DDate{Time: e.StmtTimestamp.Truncate(24 * time.Hour)}, nil
+			},
+		},
+	},
+
+	"statement_timestamp": {nowImpl},
+	"current_timestamp":   {nowImpl},
+	"now":                 {nowImpl},
```
Comment (bdarnell) [inline sql/parser/builtins.go]: In postgres, `now()` and `current_timestamp()` are equivalent to `transaction_timestamp()`, not `statement_timestamp()`.

File: sql/parser/builtins.go [hunk 12]
```diff
@@ -399,15 +401,46 @@ var builtins = map[string][]builtin{
...
+		builtin{
+			types:      typeList{},
+			returnType: DummyTimestamp,
+			fn: func(e EvalContext, args DTuple) (Datum, error) {
+				return DTimestamp{Time: e.TxnTimestamp}, nil
+			},
+		},
+	},
+
+	"transaction_date": {
```
Comment (bdarnell) [inline sql/parser/builtins.go]: Does any other database have a `transaction_date()` function? I'm having a hard time seeing a use for it and it's always expressible as `cast(transaction_timestamp() as date)`.

File: sql/parser/sql.y [hunk 13]
```diff
@@ -2807,6 +2808,14 @@ func_expr_common_subexpr:
 | GREATEST '(' expr_list ')' { unimplemented() }
 | LEAST '(' expr_list ')' { unimplemented() }
 
+func_or_variable:
```
Comment (bdarnell) [inline sql/parser/sql.y]: Only current_date and current_time have the func_or_variable treatment in postgres; the others are only functions. I'd prefer to minimize the number of names that have this special case.

File: sql/executor.go (hunk 10 above)
Comment (vivekmenezes) [inline sql/executor.go]: sounds good

File: sql/testdata/builtin_function [hunk 14]
```diff
@@ -254,16 +254,70 @@ SELECT now() - current_date::timestamp < interval '24h10s'
...
+----
+true
+
+query T
+SELECT now() - statement_timestamp = interval '0s'
+----
+true
+
+query T
+SELECT current_timestamp() - statement_timestamp() = interval '0s'
```
Comment (bdarnell) [inline sql/testdata/builtin_function]: It would be nice if we could have some tests where statement_timestamp and transaction_timestamp differ, but I don't see how to do that without introducing ugly test-only hooks.

File: sql/parser/builtins.go (hunk 11 above)
Comment (bdarnell) [inline sql/parser/builtins.go]: I don't feel strongly which way we go. It would be good, however, if functions like `current_date()` and `age()` were defined in terms of `nowImpl` to ensure that we're consistent.

File: sql/testdata/builtin_function (hunk 14 above)
Comment (petermattis) [inline sql/testdata/builtin_function]: We could add something to the logic-test framework (e.g. `sleep 10ms`).

Comment (tamird) [inline sql/testdata/builtin_function]: or we could do it in driver_test?

File: sql/parser/sql.y (hunk 13 above)
Comment (vivekmenezes) [inline sql/parser/sql.y]: I'm in favor of removing the variable treatment for all these functions. The "modernized" approach does that.

File: sql/parser/builtins.go (hunk 12 above)
Comment (vivekmenezes) [inline sql/parser/builtins.go]: I'll take it out

File: sql/parser/builtins.go (hunk 11 above)
Comment (vivekmenezes) [inline sql/parser/builtins.go]: they all share e.StmtTimestamp to ensure consistency.

File: sql/parser/sql.y (hunk 13 above)
Comment (bdarnell) [inline sql/parser/sql.y]: The variable form of `current_timestamp` is present in the SQL standard, and some databases forbid the use of parentheses with it (http://stackoverflow.com/questions/10084967/is-oracles-current-timestamp-function-really-a-function, although this information is suspect since it says that postgres disallows parens)

File: sql/executor.go [hunk 15]
```diff
@@ -54,10 +54,13 @@ func (e *Executor) SetNodeID(nodeID proto.NodeID) {
 // Execute the statement(s) in the given request and return a response.
 // On error, the returned integer is an HTTP error code.
 func (e Executor) Execute(args driver.Request) (driver.Response, int, error) {
+	txnTimestamp := time.Now()
```
Comment (tamird) [inline sql/executor.go]: this is a misnomer, this is the statement timestamp, not the transaction's.

File: rpc/codec/wire/.!8913!wire.pb.go [hunk 16]
```diff
@@ -0,0 +1,155 @@
+// Code generated by protoc-gen-gogo.
```
Comment (bdarnell) [inline rpc/codec/wire/.!8913!wire.pb.go]: You've got some sort of garbage file here.

File: sql/executor.go [hunk 17]
```diff
@@ -83,6 +88,7 @@ func (e Executor) Execute(args driver.Request) (driver.Response, int, error) {
 	if planMaker.txn != nil {
 		planMaker.session.Txn = &planMaker.txn.Proto
 		planMaker.session.MutatesSystemDB = planMaker.txn.SystemDBTrigger()
+		planMaker.session.Timestamp = driver.Timestamp(txnTimestamp)
```
Comment (tamird) [inline sql/executor.go]: I think we need to put this timestamp on the `Txn` (which might mean we need a new proto message that wraps `proto.Transaction`). it's really not session state, and this is fragile as a result; if the transaction is committed all its associated state should be blown away.

File: sql/executor.go (hunk 15 above)
Comment (bdarnell) [inline sql/executor.go]: It's both: the statement timestamp and transaction timestamp are the same by default until we read a new value for txnTimestamp from the session Transaction. But I agree it would be clearer to have two separate variables.

File: sql/parser/builtins.go (hunk 11 above)
Comment (bdarnell) [inline sql/parser/builtins.go]: Yes, they all share `e.StmtTimestamp`, but they all independently choose to use StmtTimestamp instead of TxnTimestamp. What I'm saying is that all of these should call into one place so that if we decide to make `now()` synonymous with `transaction_timestamp` there is only one place to change.

File: sql/executor.go (hunk 17 above)
Comment (bdarnell) [inline sql/executor.go]: Why is that a problem?

File: sql/parser/sql.y (hunk 13 above)
Comment (vivekmenezes) [inline sql/parser/sql.y]: we decided to kill the variable form for now since a functional form is consistent, and gels with a modern approach, and will punt changing this back to the standard variable form.

File: sql/executor.go (hunk 17 above)
Comment (tamird) [inline sql/executor.go]: as discussed:

```sql
CREATE TABLE t (a timestamp);

BEGIN; INSERT INTO t VALUES (transaction_timestamp()); COMMIT; SELECT SLEEP (1); BEGIN; SELECT transaction_timestamp(), * FROM t; COMMIT;
```

Comment (vivekmenezes) [inline sql/executor.go]: we talked and we decided to create a new message that embeds proto.Transaction and adds a sql transaction timestamp.

File: sql/executor.go (hunk 15 above)
Comment (vivekmenezes) [inline sql/executor.go]: yeah they aught to be separate and created elsewhere.

File: sql/executor.go [hunk 18]
```diff
@@ -54,10 +54,13 @@ func (e *Executor) SetNodeID(nodeID proto.NodeID) {
 // Execute the statement(s) in the given request and return a response.
 // On error, the returned integer is an HTTP error code.
 func (e Executor) Execute(args driver.Request) (driver.Response, int, error) {
+	txnTimestamp := time.Now()
 	planMaker := planner{
 		user: args.GetUser(),
 		evalCtx: parser.EvalContext{
-			NodeID: e.nodeID,
+			NodeID:        e.nodeID,
+			StmtTimestamp: txnTimestamp,
```
Comment (tamird) [inline sql/executor.go]: by the way, this suffers a similar problem to the one described below. Concretely, this will produce the wrong output for this case:

```sql
INSERT INTO t VALUES (statement_timestamp()); SELECT SLEEP (1); SELECT statement_timestamp(), * FROM t;
```

**Comment by vivekmenezes** [comment]
Hey, this PR has changed quite a bit that it's worth taking another look at it.

File: sql/testdata/datetime [hunk 19]
```diff
@@ -72,3 +82,175 @@ query B
...
+a
+
+statement ok
+COMMIT TRANSACTION
+
+# Test that the transaction_timestamp can differ from the current_timestamp.
+# Check that the transaction_timestamp changes with each transaction.
+# We use, SELECT * FROM kv, to insert delays.
+statement ok
+BEGIN; INSERT INTO kv (k,v) VALUES ('b', transaction_timestamp()); SELECT * FROM kv; INSERT INTO kv (k,v) VALUES ('c', transaction_timestamp()); SELECT * FROM kv; INSERT INTO kv (k,v) VALUES ('d', current_timestamp()); SELECT * FROM KV; COMMIT; BEGIN; SELECT * FROM KV; INSERT INTO kv (k,v) VALUES ('e', transaction_timestamp()); COMMIT;
```
Comment (petermattis) [inline sql/testdata/datetime]: FYI, this does not need to be one line. The logic-test code reads statements to execute up until the next blank line. That's how the `CREATE TABLE` at line 143 works.

File: storage/engine/.!9097!mvcc.pb.go [hunk 20]
```diff
@@ -0,0 +1,521 @@
+// Code generated by protoc-gen-gogo.
```
Comment (petermattis) [inline storage/engine/.!9097!mvcc.pb.go]: What is this file?

File: util/.!7380!unresolved_addr.pb.go [hunk 21]
```diff
@@ -0,0 +1,172 @@
+// Code generated by protoc-gen-gogo.
```
Comment (petermattis) [inline util/.!7380!unresolved_addr.pb.go]: Ditto. What is this file?

File: sql/plan.go [hunk 22]
```diff
@@ -34,6 +35,15 @@ type planner struct {
 	evalCtx parser.EvalContext
 }
 
+func (p *planner) assignTxn(txn *client.Txn, timestamp time.Time) {
```
Comment (petermattis) [inline sql/plan.go]: While this is an assignment, I think this would more typically be named `setTxn`.

File: sql/parser/sql.y [hunk 23]
```diff
@@ -3596,10 +3588,7 @@ reserved_keyword:
 | CONSTRAINT
 | CREATE
 | CURRENT_CATALOG
-| CURRENT_DATE
```
Comment (petermattis) [inline sql/parser/sql.y]: In general, you shouldn't be touching the `reserved_keyword` rule. Why was this change necessary? I'd like to see it reverted.

File: sql/parser/sql.y [hunk 24]
```diff
@@ -2774,14 +2774,6 @@ func_expr_windowless:
 // Special expressions that are considered to be functions.
 func_expr_common_subexpr:
   COLLATION FOR '(' a_expr ')' { unimplemented() }
-| CURRENT_DATE
```
Comment (petermattis) [inline sql/parser/sql.y]: I think you should revert this. The variable form of `CURRENT_{DATE,TIMESTAMP}` is part of the SQL standard.

Comment (vivekmenezes) [inline sql/parser/sql.y]: We decided to not go the standards route because it felt out of place and
is not part of the modern approach. We can always add in the standards
style later.

Comment (tamird) [inline sql/parser/sql.y]: who is we?

Comment (petermattis) [inline sql/parser/sql.y]: Huh, did I miss that discussion? I think this is a mistake. And it's not like this a burden to support. FYI, your "modern approach" information is incorrect. We need to support `CURRENT_{TIMESTAMP,DATE}` without parentheses.

File: sql/plan.go (hunk 22 above)
Comment (tamird) [inline sql/plan.go]: shouldn't `planner.txn` be a `Txn` rather than `client.Txn`?

File: sql/executor.go [hunk 25]
```diff
@@ -236,9 +234,9 @@ func (e Executor) execStmt(stmt parser.Statement, params parameters, planMaker *
 	// No transaction. Run the command as a retryable block in an
 	// auto-transaction.
 	err := e.db.Txn(func(txn *client.Txn) error {
-		planMaker.txn = txn
+		planMaker.assignTxn(txn, timestamp)
```
Comment (tamird) [inline sql/executor.go]: i don't think it's correct for this to use `timestamp` as computed earlier

File: sql/plan.go (hunk 22 above)
Comment (tamird) [inline sql/plan.go]: hm, perhaps not.

File: sql/executor.go (hunk 25 above)
Comment (vivekmenezes) [inline sql/executor.go]: Any particular reason why?

File: sql/parser/builtins.go (hunk 11 above)
Comment (vivekmenezes) [inline sql/parser/builtins.go]: They all call nowTimestamp()

File: sql/parser/sql.y (hunk 24 above)
Comment (vivekmenezes) [inline sql/parser/sql.y]: Will put them both back.

File: sql/executor.go (hunk 25 above)
Comment (tamird) [inline sql/executor.go]: well the clock continues to run - consider the case of an insert statement using transaction_timestamp: this implementation allow drift in the face of retries

Comment (vivekmenezes) [inline sql/executor.go]: Very good point. I've fixed it.

File: sql/parser/sql.y (hunk 23 above)
Comment (vivekmenezes) [inline sql/parser/sql.y]: reverted

File: sql/testdata/datetime (hunk 19 above)
Comment (vivekmenezes) [inline sql/testdata/datetime]: done

File: sql/parser/builtins.go [hunk 26]
```diff
@@ -399,15 +401,36 @@ var builtins = map[string][]builtin{
...
 	},
 
-	"current_timestamp": {nowImpl},
-	"now":               {nowImpl},
+	"transaction_timestamp": {
+		builtin{
+			types:      typeList{},
+			returnType: DummyTimestamp,
+			fn: func(e EvalContext, args DTuple) (Datum, error) {
+				return DTimestamp{Time: e.TxnTimestamp}, nil
```
Comment (tamird) [inline sql/parser/builtins.go]: use `nowTimestamp()`

Comment (tamird) [inline sql/parser/builtins.go]: err, disregard

**Comment by tamird** [comment]
LGTM
